### ISSUE TITLE
Parameter draws (stochastically) out of prior range in Python 3

### ISSUE THREAD

**Issue by jakevdp** [pymc-devs/pymc#629]
I have a very strange issue that I can't figure out – I'm running PyMC 2.3 on Python 3.3. There are two things happening:
1. Even though I explicitly set the random seed, I get different results each time I run the script
2. About 50% of the time that I run the script, the draw for the hyperparameter `Pb`, which has a uniform prior between 0 and 1, falls outside the range. If I return zero within the likelihood computation, the computation time increases from ~10s to ~90min (i.e. nearly every draw is being flat-out rejected).

Any idea what the source of the randomness is, and why this parameter is so often drawn outside its prior range? Whenever the draws are correct, I get the same result each time. Whenever the draws are incorrect, it fails essentially immediately.

I haven't been able to reproduce this in Python 2.x.

Here's a stripped-down version of the code:

``` python
import numpy as np
import pymc

print("pymc version", pymc.__version__)

np.random.seed(0)

xi = np.random.random(10)
dyi = np.random.random(10)
yi = np.random.normal(2 * xi, dyi)

# We're doing a linear fit with a background. Pb is the nuisance parameter
# which specifies the probability that points are drawn from a normal
# distribution N(Yb, sigmab^2).

# uniform prior on Pb, the fraction of bad points
Pb = pymc.Uniform('Pb', 0, 1.0, value=0.1)

# uniform prior on Yb, the centroid of the outlier distribution
Yb = pymc.Uniform('Yb', -10000, 10000, value=0)

# uniform prior on log(sigmab), the spread of the outlier distribution
log_sigmab = pymc.Uniform('log_sigmab', -10, 10, value=5)

# define priors on beta = (slope, intercept)
@pymc.stochastic
def beta_M1(value=np.array([2., 100.])):
    """Slope and intercept parameters for a straight line.
    The likelihood corresponds to the prior probability of the parameters."""
    slope, intercept = value
    prob_intercept = 1 + 0 * intercept
    # uniform prior on theta = arctan(slope)
    # d[arctan(x)]/dx = 1 / (1 + x^2)
    prob_slope = np.log(1. / (1. + slope ** 2))
    return prob_intercept + prob_slope


@pymc.deterministic
def model_M1(xi=xi, beta=beta_M1):
    slope, intercept = beta
    return slope * xi + intercept


@pymc.deterministic
def sigmab(log_sigmab=log_sigmab):
    return np.exp(log_sigmab)


# set up the expression for likelihood
def mixture_likelihood(yi, model, dyi, Pb, Yb, sigmab):
    if (Pb < 0) or (Pb > 1):
        print("error: Pb =", Pb)
        raise ValueError("Pb out of range")

    Vi = dyi ** 2
    Vb = sigmab ** 2

    root2pi = np.sqrt(2 * np.pi)

    L_in = (1. / root2pi / dyi
            * np.exp(-0.5 * (yi - model) ** 2 / Vi))

    L_out = (1. / root2pi / np.sqrt(Vi + Vb)
             * np.exp(-0.5 * (yi - Yb) ** 2 / (Vi + Vb)))

    return np.sum(np.log((1 - Pb) * L_in + Pb * L_out))

MixtureNormal = pymc.stochastic_from_dist('mixturenormal',
                                          logp=mixture_likelihood,
                                          dtype=np.float,
                                          mv=True)

y_mixture = MixtureNormal('y_mixture', model=model_M1, dyi=dyi,
                          Pb=Pb, Yb=Yb, sigmab=sigmab,
                          observed=True, value=yi)

M = dict(y_mixture=y_mixture, beta_M1=beta_M1, model_M1=model_M1,
         Pb=Pb, Yb=Yb, log_sigmab=log_sigmab, sigmab=sigmab)

S = pymc.MCMC(M)
S.sample(iter=25000, burn=5000)
```

The output when it works:

```
pymc version 2.3.3
[-----------------100%-----------------] 25000 of 25000 complete in 9.6 sec
```

The output the rest of the time:

```
pymc version 2.3.3
error: Pb = -0.8636627605010293
Traceback (most recent call last):
  File "tmp.py", line 87, in <module>
    S.sample(iter=25000, burn=5000)
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/MCMC.py", line 277, in sample
    Sampler.sample(self, iter, length, verbose)
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/Model.py", line 251, in sample
    self._loop()
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/MCMC.py", line 313, in _loop
    step_method.step()
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/StepMethods.py", line 1967, in step
    while(J > 0 and logy < self.fll(L)):
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/StepMethods.py", line 1998, in fll
    ll = self.loglike
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/StepMethods.py", line 302, in loglike
    sum = logp_of_set(self.children)
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/Node.py", line 39, in logp_of_set
    six.reraise(*exc)
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/six.py", line 619, in reraise
    raise value
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/Node.py", line 30, in logp_of_set
    logp += obj.logp
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/PyMCObjects.py", line 907, in get_logp
    logp = self._logp.get()
  File "LazyFunction.pyx", line 280, in pymc.LazyFunction.LazyFunction.get (pymc/LazyFunction.c:2568)
  File "/Users/jakevdp/anaconda/envs/py3k/lib/python3.3/site-packages/pymc/distributions.py", line 2977, in wrapper
    return f(value, **kwds)
  File "tmp.py", line 59, in mixture_likelihood
    raise ValueError("Pb out of range")
ValueError: Pb out of range
```

**Comment by twiecki** [comment]
Sounds spooky. Can you reduce this down to just sampling from a uniform?

**Comment by jakevdp** [comment]
I've tried multiple times to simplify this, but there's some magic combination in the above script that leads to the issue.

**Comment by fonnesbeck** [comment]
Try using the 2.3.4 release, or build from the 2.3 branch. If this is related to #612, then it is solved by using a more current version.

**Comment by fonnesbeck** [comment]
If this is not solved using 2.3.4, please reopen on the [PyMC 2 repo](https://github.com/pymc-devs/pymc)
